Fix issue with FloorDiv handling

nandesuka opened the pull request:
Summary:
The launch grid calculation is generated as string and evaluated to get sympy expressions. 

The default "python" mode generates expressions using negative integer division behaviour in python. This creates problems when we try generate generic code for representing the launch grid calculation at the FX graph level.

In this diff we:

1. Add a new mode "python_slow" which achieves the same ceildiv behaviour that is generally applicable to integer division
2. Move and generalize the logic for replacing sympy.floor to CeilDiv

Test Plan:
CI

Rollback Plan:

Differential Revision: D80744150




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

blaine-rister commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -2980,14 +2980,14 @@ class GridExpr:
 
     inductor_meta: dict[str, Any]
-    mode: Literal["python", "cpp"] = "python"
+    mode: Literal["python", "cpp", "python_slow"] = "python"
```
Is `"python_slow"` mode used by the FX converter? I'm trying to figure out where that's configured.

blaine-rister reviewed (approved):
LGTM!

nandesuka committed:
Support generic dynamic shape with padding (#160997)

Summary:

Inductor has the following configurations:

config.comprehensive_padding
config.padding_alignment_bytes
config.padding_stride_threshold

In the case of static shape by enabling these three options Inductor will generate code for Flexible layout tensors that tries to pad up all stride dimension to be a multiple of config.padding_alignment_bytes for strides above: config.padding_stride_threshold. In the case where dynamic shapes is enabled no padding is done today.

This PR introduces the following configuration which allows the user to specify they wish to generated a padded stride even in the case of dynamic shape operations. This is mainly done so we don't break the previous behaviour of not padding up dynamic shape use cases. The config.padding_stride_threshold does not apply since the values of the strides are dynamic.

config.pad_dynamic_shapes

Test Plan:
CI

Rollback Plan:

Reviewed By: blaine-rister

Differential Revision: D80468808

nandesuka committed:
Fix issue with FloorDiv handling (#161410)

Summary:

The launch grid calculation is generated as string and evaluated to get sympy expressions. 

The default "python" mode generates expressions using negative integer division behaviour in python. This creates problems when we try generate generic code for representing the launch grid calculation at the FX graph level.

In this diff we:

Add a new mode "python_slow" which achieves the same ceildiv behaviour that is generally applicable to integer division

Test Plan:
CI

Rollback Plan:

Reviewed By: blaine-rister

Differential Revision: D80744150
